base/src/MultimediaQueue.cpp

```diff
@@ -0,0 +1,789 @@
+			BOOST_LOG_TRIVIAL(info) << "I-FRAME" << typeFound;
+
+			if (spsBuff.size() != 0)
+			{
+				m_headerFrame = frame;
+				spsBuffer = spsBuff;
+				ppsBuffer = ppsBuff;
+			}
+			mQueue.insert({ it->second->timestamp, frames });
+
+			if (largestTimeStamp < it->second->timestamp)
+			{
+				largestTimeStamp = it->second->timestamp;
+			}
+		}
+		BOOST_LOG_TRIVIAL(info) << "queue size = " << mQueue.size();
+		if (isMapDelayInTime) // If the lower and upper watermark are given in time
+		{
+			if ((largestTimeStamp - mQueue.begin()->first > lowerWaterMark) && (pushNext == true))
+			{
+				mQueue.erase(mQueue.begin()->first);
+				auto it = mQueue.begin();
+				while (it != mQueue.end())
+				{
+					auto itr = it;
+					++it;
+					auto frame = itr->second.begin()->second;
+					mutable_buffer& h264Frame = *(frame.get());
+					auto ret = H264Utils::parseNalu(h264Frame);
+					tie(typeFound, inFrame, spsBuff, ppsBuff) = ret;
+
+					if (typeFound == H264Utils::H264_NAL_TYPE::H264_NAL_TYPE_IDR_SLICE)
+					{
+						break;
+					}
+					mQueue.erase(itr->first);
+				}
+
+			}
+
+			else if ((largestTimeStamp - mQueue.begin()->first > upperWaterMark) && (pushNext == false))
+			{
+				auto lastElement = mQueue.end();
+				lastElement--;
+				auto lastElementTimeStamp = lastElement->first;
+				mQueue.erase(mQueue.begin()->first);
+				auto it = mQueue.begin();
+				while (it != mQueue.end())
+				{
+
+					auto itr = it;
+					++it;
+					auto frame = itr->second.begin()->second;
+					mutable_buffer& h264Frame = *(frame.get());
+					auto ret = H264Utils::parseNalu(h264Frame);
+					tie(typeFound, inFrame, spsBuff, ppsBuff) = ret;
+
+					if (typeFound == H264Utils::H264_NAL_TYPE::H264_NAL_TYPE_SEQ_PARAM)
+					{
+						break;
+					}
+					mQueue.erase(itr->first);
+
+					if ((lastElementTimeStamp - it->first) < lowerWaterMark)
+					{
+						break;
+					}
+
+				}
+				pushNext = true;
+			};
+		}
+
+		else // If the lower and upper water mark are given in number of frames
+		{
+			if ((mQueue.size() > lowerWaterMark) && (pushNext == true))
+			{
+				mQueue.erase(mQueue.begin()->first);
+			}
+
+			if ((mQueue.size() > upperWaterMark) && (pushNext == false))
+			{
+				auto it = mQueue.begin();
+				while (it != mQueue.end())
+				{
+					if (mQueue.size() < lowerWaterMark)
+					{
+						break;
+					}
+					auto itr = it;
+					++it;
+					auto frame = it->second.begin()->second;
+					mutable_buffer& h264Frame = *(frame.get());
+					auto ret = H264Utils::parseNalu(h264Frame);
+					tie(typeFound, inFrame, spsBuff, ppsBuff) = ret;
+
+					if (typeFound == H264Utils::H264_NAL_TYPE::H264_NAL_TYPE_IDR_SLICE)
+					{
+						break;
+					}
+					mQueue.erase(itr->first);
+				}
+				pushNext = true;
+			};
+		}
+		return true;
+	}
+protected:
+	frame_sp m_headerFrame;
+	const_buffer spsBuff;
+	const_buffer ppsBuff;
+	const_buffer inFrame;
+	short typeFound;
+};
+
+//State Design begins here
+class Export : public State {
```

Comment:
Export is too generic it should be called ExportQState and same applies to others